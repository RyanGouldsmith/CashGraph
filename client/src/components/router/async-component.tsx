import React from 'react';

interface DynamicComponentInterface {
  [key: string]: React.ComponentClass<{}, {}> | React.SFC<{}>;
}

interface ComponentState {
  Component: null | React.ComponentClass<{}, {}> | React.SFC<{}>;
}

export const AsyncComponent = (
  loader: () => Promise<DynamicComponentInterface>,
  componentName: string,
) =>
  class AsyncComp extends React.Component<{}, ComponentState> {
    constuctor() {
      this.state = {
        Component: null,
      };
    }

    public async componentDidMount() {
      const ComponentItem = await loader();
      this.setState({ Component: ComponentItem[componentName] });
    }

    public render() {
      const Component = this.state && this.state.Component;
      return Component ? <Component {...this.props} /> : null;
    }
  };
