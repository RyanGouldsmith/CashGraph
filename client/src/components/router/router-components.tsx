import React from 'react';

export const Dashboard = React.lazy(() =>
  import(/* webpackChunkName: "dashboard" */ '../dashboard/dashboard'));

export const Spending = React.lazy(() =>
  import(/* webpackChunkName: "spending" */ '../spending/spending'));

export const SpendingEdit = React.lazy(() =>
  import(/* webpackChunkName: "spending-edit" */ '../spending/edit-spending'));

export const SpendingDelete = React.lazy(() =>
  import(/* webpackChunkName: "spending-delete" */ '../spending/delete-spending'));

export const Error = React.lazy(() => import(/* webpackChunkName: "error" */ '../error/error'));

export const CreateSpending = React.lazy(() =>
  import(/* webpackChunkName: "create-spending" */ '../spending/create-spending'));

export const EditUser = React.lazy(() =>
  import(/* webpackChunkName: "edit-user" */ '../user/edit-user/edit-user'));
