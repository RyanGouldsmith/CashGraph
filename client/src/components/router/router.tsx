import React from 'react';
import { Switch, Route } from 'react-router-dom';
import { Loading } from '../loading/loading';

const Dashboard = React.lazy(() =>
  import(/* webpackChunkName: "dashboard" */ '../dashboard/dashboard'),
);

const Spending = React.lazy(() =>
  import(/* webpackChunkName: "spending" */ '../spending/spending'),
);

const Error = React.lazy(() => import(/* webpackChunkName: "error" */ '../error/error'));

const CreateSpending = React.lazy(() =>
  import(/* webpackChunkName: "create-spending" */ '../spending/create-spending'),
);

export const Router: React.SFC<{}> = _ => {
  return (
    <main className="content">
      <React.Suspense fallback={<Loading />}>
        <Switch>
          <Route exact path="/" component={Dashboard} />
          <Route exact path="/spending/create" component={CreateSpending} />
          <Route exact path="/spending" component={Spending} />
          <Route component={Error} />
        </Switch>
      </React.Suspense>
    </main>
  );
};
