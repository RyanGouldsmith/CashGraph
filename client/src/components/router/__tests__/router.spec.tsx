import 'babel-polyfill';
import React from 'react';
import { Router } from '../router';
import { render, wait, cleanup } from 'react-testing-library';
import { MockedProvider } from 'react-apollo/test-utils';
import { MemoryRouter } from 'react-router';

afterEach(cleanup);

test('Router should not render any child component for a non-existant path', async () => {
  const routerOutput = render(
    <MockedProvider mocks={[]}>
      <MemoryRouter initialEntries={['/no-go']}>
        <Router />
      </MemoryRouter>
    </MockedProvider>,
  );

  await wait(() => {
    expect(routerOutput).toMatchSnapshot();
  });
});

test('Router should render the dashboard component for path `/`', async () => {
  const routerOutput = render(
    <MockedProvider mocks={[]}>
      <MemoryRouter initialEntries={['/']}>
        <Router />
      </MemoryRouter>
    </MockedProvider>,
  );

  await wait(() => {
    expect(routerOutput).toMatchSnapshot();
  });
});

test('Router should render the Spending component for path `/spending`', async () => {
  const routerOutput = render(
    <MockedProvider mocks={[]}>
      <MemoryRouter initialEntries={['/spending']}>
        <Router />
      </MemoryRouter>
    </MockedProvider>,
  );

  await wait(() => {
    expect(routerOutput).toMatchSnapshot();
  });
});

test('Router should render the CreateSpending component for path `/spending/create`', async () => {
  const routerOutput = render(
    <MockedProvider mocks={[]}>
      <MemoryRouter initialEntries={['/spending/create']}>
        <Router />
      </MemoryRouter>
    </MockedProvider>,
  );

  await wait(() => {
    expect(routerOutput).toMatchSnapshot();
  });
});

test('Router should render the EditUser component for path `/user/edit`', async () => {
  const routerOutput = render(
    <MockedProvider mocks={[]}>
      <MemoryRouter initialEntries={['/user/edit']}>
        <Router />
      </MemoryRouter>
    </MockedProvider>,
  );

  await wait(() => {
    expect(routerOutput).toMatchSnapshot();
  });
});

test('Router should render the EditSpending component for path `/spending/edit/123`', async () => {
  const routerOutput = render(
    <MockedProvider mocks={[]}>
      <MemoryRouter initialEntries={['/spending/edit/123']}>
        <Router />
      </MemoryRouter>
    </MockedProvider>,
  );

  await wait(() => {
    expect(routerOutput).toMatchSnapshot();
  });
});
