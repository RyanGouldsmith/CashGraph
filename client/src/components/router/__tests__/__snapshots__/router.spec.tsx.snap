// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router should not render any child component for a non-existant path 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="content"
      >
        <h1>
          Sorry! That page does not exist
        </h1>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="content"
    >
      <h1>
        Sorry! That page does not exist
      </h1>
    </main>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Router should render the CreateSpending component for path \`/spending/create\` 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="content"
      >
        <form>
          <label
            for="title"
          >
            Enter the item you purchased
          </label>
          <input
            id="title"
            placeholder="Shoes"
            type="text"
          />
          <label
            for="price"
          >
            The price of the item
          </label>
          <input
            id="price"
            min="0"
            type="number"
          />
          <label
            for="tag"
          >
            Type of Tag
          </label>
          <input
            id="tag"
          />
          <label
            for="tag-colour"
          >
            Tag Colour
          </label>
          <input
            id="tag-colour"
          />
          <button
            type="submit"
          >
            submit
          </button>
        </form>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="content"
    >
      <form>
        <label
          for="title"
        >
          Enter the item you purchased
        </label>
        <input
          id="title"
          placeholder="Shoes"
          type="text"
        />
        <label
          for="price"
        >
          The price of the item
        </label>
        <input
          id="price"
          min="0"
          type="number"
        />
        <label
          for="tag"
        >
          Type of Tag
        </label>
        <input
          id="tag"
        />
        <label
          for="tag-colour"
        >
          Tag Colour
        </label>
        <input
          id="tag-colour"
        />
        <button
          type="submit"
        >
          submit
        </button>
      </form>
    </main>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Router should render the Spending component for path \`/spending\` 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="content"
      >
        <h1>
          Spending ... 
        </h1>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="content"
    >
      <h1>
        Spending ... 
      </h1>
    </main>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Router should render the dashboard component for path \`/\` 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="content"
      >
        <h1>
          Cash Graph Dashboard
        </h1>
        <main
          class="dashboard"
        >
          <section
            class="dashboard__tags"
          >
            <h1>
              The tags are: 
            </h1>
          </section>
          <section
            class="dashboard__spending"
          >
            <a
              href="/spending"
            >
              More Spending ...
            </a>
            <h1>
              Spending ... 
            </h1>
          </section>
        </main>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="content"
    >
      <h1>
        Cash Graph Dashboard
      </h1>
      <main
        class="dashboard"
      >
        <section
          class="dashboard__tags"
        >
          <h1>
            The tags are: 
          </h1>
        </section>
        <section
          class="dashboard__spending"
        >
          <a
            href="/spending"
          >
            More Spending ...
          </a>
          <h1>
            Spending ... 
          </h1>
        </section>
      </main>
    </main>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
